services/MerchStoreService.ts

```diff
@@ -100,32 +102,49 @@ export default class MerchStoreService {
     const item = await this.transactions.readOnly(async (txn) => Repositories
       .merchStoreItem(txn)
       .findByUuid(uuid));
-    if (!item) throw new NotFoundError('Item not found');
+    if (!item) throw new NotFoundError(NotFoundErrors.MERCH_ITEM);
     return item.getPublicMerchItem();
   }
 
   public async createItem(item: MerchItem): Promise<MerchandiseItemModel> {
     return this.transactions.readWrite(async (txn) => {
+      if (this.hasMultipleOptionTypes(item.options)) throw new UserError(UserErrors.MULTIPLE_MERCH_OPTION_TYPES);
+      if (!item.hasVariants && item.options.length > 1) {
+        throw new UserError(UserErrors.NO_ITEM_VARIANTS_BUT_MULTIPLE_OPTIONS);
+      }
+
       const collection = await Repositories.merchStoreCollection(txn).findByUuid(item.collection);
-      if (!collection) throw new NotFoundError('Collection not found');
+      if (!collection) throw new NotFoundError(NotFoundErrors.MERCH_COLLECTION);
+
       const merchItemRepository = Repositories.merchStoreItem(txn);
       const merchItem = MerchandiseItemModel.create({ ...item, collection });
       await merchItemRepository.upsertMerchItem(merchItem);
       return merchItemRepository.findByUuid(merchItem.uuid);
     });
   }
 
+  private hasMultipleOptionTypes(options: MerchItemOption[] | MerchandiseItemOptionModel[] | MerchItemOptionEdit[]) {
+    for (let i = 1; i < options.length; i += 1) {
+      if (options[i].metadata?.type !== options[i - 1].metadata?.type) {
+        return true;
+      }
+    }
+    return false;
+  }
+
   public async editItem(uuid: Uuid, itemEdit: MerchItemEdit): Promise<MerchandiseItemModel> {
     return this.transactions.readWrite(async (txn) => {
       const merchItemRepository = Repositories.merchStoreItem(txn);
       const item = await merchItemRepository.findByUuid(uuid);
       if (!item) throw new NotFoundError();
-      const { options, collection: updatedCollection, ...changes } = itemEdit;
+      const { options, collection: updatedCollection, hasVariants, ...changes } = itemEdit;
+      if (this.hasMultipleOptionTypes(options)) throw new UserError(UserErrors.MULTIPLE_MERCH_OPTION_TYPES);
```

Comment:
What happens if the options in the edit all have some type and all the options in the model have some other type?

Reply:
Great point - there's 2 approaches I'm thinking for that:
1. Keep the check on line 141. After that, check if `itemEdit.options.length` is equal to `item.options.length` (i.e. every option is being edited) and throw if the `type` field is changed in the edit. If not every option is being edited and instead the `itemEdit.options.length` array is filled with bogus edits that make the above length check pass, then the `Option Not Found` error will be thrown instead in the code below
2. Remove the check on line 141, and after [this](https://github.com/acmucsd/membership-portal/blob/4a9d27d7f90f90d953d8d0111c452f27d1c03dff/services/MerchStoreService.ts#L137) line, we do:
```ts
item.options = [...updatedOptions, ...item.options.filter((option) => !updatedOptionsIds.has(option.uuid))];
if (this.hasMultipleOptionTypes(item.options)) throw new UserError(UserErrors.MULTIPLE_MERCH_OPTION_TYPES);
```

Approach 2 is a lot simpler but the only downside is reads/writes are potentially wasted in the `Promise.all` call, but since the service is transacted, we can just throw and everything will get reverted. But again it doesn't really make sense to write data and then validate after it's been written. Meanwhile, approach 1 doesn't waste reads/writes but is a bit more complex of a check. Thoughts / is there any other approach you're thinking could work?

Reply:
Why would we need to write before validating? We already have the current item and its options, we merge the edits, we validate the updated item, and then we write the updated item.